lib/Conversion/SecretToCKKS/SecretToCKKS.cpp

```diff
@@ -0,0 +1,157 @@
+  auto result = ::mlir::polynomial::IntPolynomial::fromMonomials(monomials);
+  if (failed(result)) return failure();
+  ::mlir::polynomial::IntPolynomial xnPlusOne = result.value();
+  switch (coefficientModBits) {
+    case 29: {
+      auto type = IntegerType::get(ctx, 32);
+      return ::mlir::polynomial::RingAttr::get(
+          type, IntegerAttr::get(type, APInt(32, 463187969)),
+          polynomial::IntPolynomialAttr::get(ctx, xnPlusOne));
+    }
+    default:
+      return failure();
+  }
+}
+
+}  // namespace
+
+// Remove this class if no type conversions are necessary
+class SecretToCKKSTypeConverter : public TypeConverter {
+ public:
+  SecretToCKKSTypeConverter(MLIRContext *ctx,
+                            ::mlir::polynomial::RingAttr rlweRing) {
+    addConversion([](Type type) { return type; });
+
+    // Convert secret types to LWE ciphertext types
+    addConversion([ctx, this](secret::SecretType type) -> Type {
+      int bitWidth =
+          llvm::TypeSwitch<Type, int>(type.getValueType())
+              .Case<RankedTensorType>(
+                  [&](auto ty) -> int { return ty.getElementTypeBitWidth(); })
+              .Case<IntegerType>([&](auto ty) -> int { return ty.getWidth(); });
+      return lwe::RLWECiphertextType::get(
+          ctx,
+          lwe::PolynomialEvaluationEncodingAttr::get(ctx, bitWidth, bitWidth),
+          lwe::RLWEParamsAttr::get(ctx, 2, ring_), type.getValueType());
+    });
+
+    ring_ = rlweRing;
+  }
+
+ private:
+  ::mlir::polynomial::RingAttr ring_;
+};
+
+struct SecretToCKKS : public impl::SecretToCKKSBase<SecretToCKKS> {
+  using SecretToCKKSBase::SecretToCKKSBase;
+
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    auto *module = getOperation();
+
+    auto rlweRing = getRlweRing(context, coefficientModBits, polyModDegree);
+    if (failed(rlweRing)) {
+      return signalPassFailure();
+    }
+    // Ensure that all secret types are uniform and matching the ring
+    // parameter size.
+    WalkResult compatibleTensors = module->walk([&](Operation *op) {
+      for (auto value : op->getOperands()) {
+        if (auto secretTy = dyn_cast<secret::SecretType>(value.getType())) {
+          auto tensorTy = dyn_cast<RankedTensorType>(secretTy.getValueType());
+          if (tensorTy && tensorTy.getShape() !=
+                              ArrayRef<int64_t>{rlweRing.value()
+                                                    .getPolynomialModulus()
+                                                    .getPolynomial()
+                                                    .getDegree()}) {
+            return WalkResult::interrupt();
+          }
+        }
+      }
+      return WalkResult::advance();
+    });
+    if (compatibleTensors.wasInterrupted()) {
+      module->emitError(
+          "expected batched secret types to be tensors with dimension "
+          "matching ring parameter");
+      return signalPassFailure();
```

Comment:
In our last meeting we had briefly discussed replication for values that can divide into 1024. I know at some point in this lowering we should be adding some way to replicate these values. Should we be adding it into this PR or in the future?

Reply:
I think that would go in a pre-pass to separate that functionality - align-tensor-sizes would work out of the box if we have everything in power of two form for it since it just replicates into the given tensor size.